### ISSUE TITLE
Automatic indentation behavior

### ISSUE THREAD

**Issue by Krzysztof-Cieslak** [Microsoft/vscode#19303]
Automatic indentation currently coming out of the box with VSCode is really unhelpful in any white-space significant programming language.

Twitter discussion (F#): https://twitter.com/isaac_abraham/status/824035064807731201
Problem reported in Python plugin: https://github.com/DonJayamanne/pythonVSCode/issues/666
Example behavior in F# (notice how new errors are shown after this automatic formatting):
![ident](https://cloud.githubusercontent.com/assets/5427083/22271801/c3790110-e296-11e6-8b0f-cc46d874471a.gif)

As far as I understand this behavior can be controlled using `IndentationRules` in `LanguageConfiguration` but:
1. They are RegEx based which won't allow us to create accurate behavior.
2. Default behaviour is completely wrong on any white-space significant language.

I think it would be good to add option to disable this behavior.

**Comment by isaacabraham** [comment]
"good option", more like "must have". This is driving me crazy.

**Comment by forki** [comment]
![brackets](https://cloud.githubusercontent.com/assets/57396/22284785/133dda16-e2e9-11e6-9ab0-2cb5cde1b156.gif)

^^ shows 2 ugly things in one

**Comment by mousetraps** [comment]
Oh no! Looks frustrating - sorry about that! @rebornix and @alexandrudima to the rescue!

**Comment by alexandrudima** [comment]
Fixed via b1c8dad73e56c5896fe5ce102eebb66aec9a47f0

**Comment by forki** [comment]
wait a second. Did you add a fix for the concrete issues or did you allow us to disable it completely? Because I think we really need the second one.

**Comment by alexandrudima** [comment]
That fix targets a different issue similar in nature, and the fix is that typing a closing bracket will not change the current line's indentation if the bracket is preceded by any non-whitespace characters, covering both issues exemplified in the two gifs in this thread.

Tomorrow's insiders build will contain the change and it would be great if you could try it out and provide additional feedback. We can then proceed to reopen this issue or start a new one, as you prefer.

**Comment by isaacabraham** [comment]
@alexandrudima I really think having the ability to turn the auto-format off is important. Not everyone wants it.

**Comment by forki** [comment]
it makes many assumptions about regularity and syntax in languages. Bad if
your language has different options.

Am 25.01.2017 9:45 nachm. schrieb "Isaac Abraham" <notifications@github.com

**Comment by forki** [comment]
![brackets](https://cloud.githubusercontent.com/assets/57396/22309805/2044542e-e34d-11e6-9ed3-c10ebb49f523.gif)

there are millions of new edge cases to find...

**Comment by rebornix** [comment]
@forki are you using the Stable version? The one you just mentioned is fixed in Insider.

I can see why you want to have the option to turn off this feature. Right now when there is an indentation adjustment, it's triggered by one of the following:

* EnterRules
* Electric Char
* Indentation Rules
* Extensions ...

The first two were implemented first to provide a reasonable experience with brackets and later we introduced Indentation Rules as other editors do. Indentation rules are not silver bullet, they can't make the indentation perfect as they are just talking about limited number of lines around.

Our target is not making indentation rules perfect as that's not possible but we want to make it reasonable. For users who want to make the indentation correct always, a proper formatter which reads AST can actually help, for example `gfmt` for Go. But there are a large amount of users who don't want to have the formatters as they are too heavy, they may just want the editor to provide a reasonable indentation but nothing else.

This iteration we enhanced Format with `formatOnPaste` and improved regex based indentation rules at the same time. We have options for Format so of course it's **a good idea** to have options for Auto Indentation. Like I said at the beginning, it's a valid feature request.

For anyone who is interested in this area, you may want to take a look at https://github.com/Microsoft/vscode/issues/17868 and PRz are always welcome.

**Comment by forki** [comment]
Yes I'm on the released bits. Regarding indentation: just moving everything 4 (or 2 depending on project) chars to the right on tab would be the 100% correct solution for us. No need for ast or any other heuristic.

**Comment by alexandrudima** [comment]
@forki @isaacabraham @Krzysztof-Cieslak 

Can you please try out the latest insider that went out a few hours ago? And provide concrete examples of what is not working as expected. Thank you!

**Comment by isaacabraham** [comment]
Here's one. 
![test](https://cloud.githubusercontent.com/assets/1781813/22325691/37c0768c-e3a7-11e6-9666-0135bc2badb4.gif)

**Comment by isaacabraham** [comment]
Another similar one: -
![test](https://cloud.githubusercontent.com/assets/1781813/22325736/68d3a0a0-e3a7-11e6-91b8-c1a2997fe87f.gif)

**Comment by isaacabraham** [comment]
I can confirm that some of the other cases that I see on the mainline build are fixed on the insider build. But I still think you should allow people to turn this off - people might use a formatting style you don't cater for, or simply not want it at all.

**Comment by gandhis1** [comment]
Let me know if this is unrelated, but I see a similar issue with Python as I reported in https://github.com/DonJayamanne/pythonVSCode/issues/666

![image](https://cloud.githubusercontent.com/assets/21064086/22371973/5e7dddc0-e468-11e6-802a-0957c128d42c.png)

**Comment by forki** [comment]
I also noticed that vscode sometimes  is inserting braces we I press space.
Urgs

Am 27.01.2017 2:13 nachm. schrieb "Siddhartha Gandhi" <
notifications@github.com>:

**Comment by gandhis1** [comment]
I tested this on the latest inside, and it appeared to be fixed. 

Although my question would be now, is when you have an open bracket expression that has yet to be closed, and you press enter, why the default behavior is not to indent to the parenthesis location.

**Comment by peredin** [comment]
I definitely want to be able to disable any automatic indentation completely.

**Comment by rebornix** [comment]
`editor.autoIndent` setting is now enriched to allow you disable this feature completely.
